### PR TITLE
Trigger snapshot at end of recovery

### LINKED ISSUE #3665: Force snapshot on recovery completion
As discussed with @wintersteiger, now that we have #3624, it is possible that a recovered service (`N+1`, that started from a snapshot emitted by the previous service `N`) does not process a sufficient number of transactions of generate a new snapshot. If the service `N+1` crashes, then the new service `N+2`, that starts from the latest generated snapshot emitted by service `N`, will not be able to verify the snapshot if the `previous_service_cert_file` specified by the operator on startup if the one of the latest service, i.e. `N+1`.

To reduce the possibility of this happening, we should force the generation of a snapshot when the private recovery is complete. This would mean that the service `N+2` would start from the latest generated snapshot, created by service `N+1`.

In fact, we should probably also force the generation of a snapshot when the initial service is open.

### PR DESCRIPTION
**Pull request by wintersteiger**
Fixes #3665

### CODE REVIEW

**Review by achamayou — APPROVED** [review]

**Comment by achamayou** [comment]
Although the change looks reasonable at first glance, it seems to have a massive impact on memory consumption!

File: src/node/node_state.h [hunk 1]
```diff
@@ -1372,6 +1377,7 @@ namespace ccf
 
         GenesisGenerator g(network, tx);
         g.open_service();
+        trigger_snapshot(tx);
```
Comment (eddyashton) [inline src/node/node_state.h:1380]: I'm guessing this is the perf hit. This adds a snapshot on every service open, and the perf tests probably didn't produce any snapshot previously.

Comment (wintersteiger) [reply src/node/node_state.h:1380]: Yeah, that's a good point! I would have expected that impact to be less severe, but you may be right.

**Comment by wintersteiger** [comment]
Not that the logic is correct, it looks like the performance also improved to a level where I'm happy with it.

**Pull request merged**